draft-ietf-tls-ticketrequests.md

```diff
@@ -124,32 +152,74 @@ Clients MAY send this extension in ClientHello. It contains the following struct
+negotiate a fresh session (full handshake), thereby implicitly invalidating all
+other tickets sharing the same initial session as the presented ticket.  This
+is also the ticket count requested when the client presented no initial ticket
+making a full handshake unavoidable.
+
+resumption_count
+: The number of tickets desired by the client when the server is willing to
+resume the associated session.
+
+Clients can use the above structure to indicate the number of tickets they
+would prefer to receive when sessions are resumed or new sessions are
+negotiated.
+
+Typically, once a client's ticket cache is primed, a resumption count of 1 is a
+good choice that allows the server to replace each ticket with a fresh ticket,
+without over-provisioning the client with excess tickets.  However, clients that
+are racing multiple connections placing a separate ticket in each, will
+ultimately end up with just the tickets from a single resumed session, so
+in that case a resumption_count commensurate with the number of parallel
+sessions would be used.
+
+On the other hand, with e.g. server-to-server traffic with fixed source
+addresses and no connection racing, ticket reuse may be appropriate.  Clients
+that wish to reuse tickets when possible should send an extension with the
+resumption_count set to 0, and the new_session_count set to the value desired
+for a full handshake (typically 1).
+
+Servers SHOULD NOT send more tickets than requested for the handshake type
+selected by the server (resumption or full handshake) as clients will most
+likely discard any additional tickets.
+
+When a resumption is performed, and the requested resumption_count is zero, the
+client is requesting ticket reuse.  If the presented ticket is valid for
+additional resumptions (only possible when the server supports reuse), it
+SHOULD return no tickets.  Otherwise, when either the server does not support
+reuse, or the ticket is expiring, and needs to be replaced (e.g. reissued under
+a fresh session-ticket encryption key) the server SHOULD return one ticket.
+
+Servers SHOULD additionally place a limit on the number of tickets they are
+willing to send, to save resources.  Therefore, the number of NewSessionTicket
+messages sent will typically be the minimum of the server's self-imposed limit
+and the number requested.
+
+A server that supports ticket requests MAY echo the "ticket_request" extension
+in the EncryptedExtensions message. If present, it contains a
+TicketRequestContents structure, where TicketRequestContents.new_session_count
+indicates the number of tickets the server expects to send to the client.
+
+If the extension is echoed, the TicketRequestContents.resumption_count can
+be set to a non-zero value to indicate that the server supports ticket
+reuse, otherwise (i.e. zero) the server does not support ticket reuse.  Ticket
+reuse has privacy implications and should be used with care or not at all,
+see the security considerations section.
```

Comment:
Yes, if we could rely on servers that support reuse to always explicitly signal that fact in an echoed extension, then the 0->1 magic would not be needed, since clients would never attempt reuse with a server that failed to promise it on the initial handshake.

It makes the client's state much more complicated (having to remember how the ticket was acquired, ...), especially in multi-process shared caches, ...  But it is not impossible.

IMHO the 0->1 thing is much cheaper, and the echoed extension is then not really needed. This could use some discussion.

Reply:
I mentioned it on the list, so we should probably have that discussion on the list.  (Not that it sounded like you were planning to have it here anyway...)